assets/src/edit-story/components/form/dropDown/list.js

```diff
@@ -0,0 +1,222 @@
+`;
+
+const availableKeysForSearch = 'abcdefghijklmnopqrstuvwxyz0123456789'.split('');
+
+function DropDownList({
+  handleCurrentValue,
+  value,
+  ariaLabel,
+  options,
+  toggleOptions,
+}) {
+  const listContainerRef = useRef();
+  const arrayOfOptionsRefs = [];
+
+  const [searchValue, setSearchValue] = useState('');
+  const [focusedValue, setFocusedValue] = useState(value);
+  const isNullOrUndefined = (item) => item === null || item === undefined;
+  const focusedIndex = useMemo(
+    () =>
+      options.findIndex(
+        (item) =>
+          !isNullOrUndefined(focusedValue) &&
+          item.value.toString() === focusedValue.toString()
+      ),
+    [focusedValue, options]
+  );
+
+  const handleMoveFocus = useCallback(
+    (offset) => {
+      if (
+        focusedIndex + offset >= 0 &&
+        focusedIndex + offset < options.length
+      ) {
+        setFocusedValue(options[focusedIndex + offset].value);
+      }
+    },
+    [focusedIndex, options]
+  );
+
+  const handleUpDown = useCallback(
+    ({ key }) => {
+      if (key === 'ArrowUp' && focusedIndex !== 0) {
+        handleMoveFocus(-1);
+      } else if (key === 'ArrowDown' && focusedIndex < options.length - 1) {
+        handleMoveFocus(1);
+      }
+    },
+    [focusedIndex, options, handleMoveFocus]
+  );
+
+  const [clearSearchValue] = useDebouncedCallback(() => {
+    setSearchValue('');
+  }, 800);
+
+  const handleKeyDown = useCallback(
+    ({ keyCode }) => {
+      const searchTerm = searchValue + String.fromCharCode(keyCode);
+      setSearchValue(searchTerm);
+      const searchIndex = options.findIndex((item) =>
+        item.name.toLowerCase().startsWith(searchTerm.toLowerCase())
+      );
+      if (searchIndex >= 0) {
+        setFocusedValue(options[searchIndex].value);
+      }
+      clearSearchValue();
+    },
+    [clearSearchValue, searchValue, options]
+  );
+
+  const handleEnter = useCallback(() => {
+    handleCurrentValue(focusedValue);
+  }, [focusedValue, handleCurrentValue]);
+
+  useFocusOut(listContainerRef, toggleOptions);
+
+  useKeyDownEffect(listContainerRef, { key: 'esc' }, toggleOptions, [
+    toggleOptions,
+  ]);
+  useKeyDownEffect(
+    listContainerRef,
+    { key: ['up', 'down'], shift: true },
+    handleUpDown,
+    [handleUpDown]
+  );
+  useKeyDownEffect(
+    listContainerRef,
+    { key: availableKeysForSearch, shift: true },
+    handleKeyDown,
+    [handleKeyDown]
+  );
+  useKeyDownEffect(
+    listContainerRef,
+    { key: ['space', 'enter'], shift: true },
+    handleEnter,
+    [handleEnter]
+  );
+
+  useEffect(() => {
+    if (!isNullOrUndefined(focusedValue)) {
+      if (focusedIndex < 0) return;
+      arrayOfOptionsRefs[focusedIndex].focus();
+    }
+  }, [focusedValue, options, focusedIndex, arrayOfOptionsRefs]);
+
+  const handleItemClick = (option) => {
+    handleCurrentValue(option);
+  };
+
+  const setOptionRef = (element) => {
+    if (element !== null) {
+      arrayOfOptionsRefs.push(element);
```

Comment:
This will be pushing the same element over and over into the array. At the very least we should dedupe this array. But best yet we should see if this can be done some other way. And the order of this array is likely not stable either. I'm also noticing that you have IDs available there and maybe that will be useful. But on a very basic level, you couldl always do `listRef.current.children[index]`.

Reply:
Yes, 🤦‍♂ Actually bad practice. Updated using listRef on parent level of listitems.